doc/source/rllib/rllib-fault-tolerance.rst

```diff
@@ -0,0 +1,74 @@
+.. include:: /_includes/rllib/announcement.rst
+
+.. include:: /_includes/rllib/we_are_hiring.rst
+
+Fault Tolerance And Elastic Training
+======================================================
+
+RLlib handles common failures modes (e.g. machine failures, spot isntance preemption,
+network outages, or Ray cluster failures) via logics at multiple system levels.
+This doc discusses them in details.
+
+Fault Tolerance and Recovery Provided by Ray Tune
+-------------------------------------------------
+
+Ray Tune provides fault tolerance and recovery at trial level.
+
+Majority of RLlib runs are launched with Ray Tune, these trials periodically checkpoint
+the entire training states to user specified persistent storage locations. If a trial fails,
+Ray Tune will automatically restart it from the latest
+:ref:`checkpointed <tune-two-types-of-ckpt>` state.
+
+This serves as our last line of defense, catching any unexpected errors that are not
+handled by RLlib.
+
+Elastic RolloutWorkers
+----------------------
+
+RLlib supports self-recovering and elastic workersets for both rollout and evaluation workers.
+This provides fault tolerance at worker level.
+
+Worker fault tolerance can be turned on by setting config ``recreate_failed_workers`` to True.
+Under the hood, RLlib relies on Ray Core :ref:`actor fault tolerance <actor-fault-tolerance>`
+to automatically recover failed worker actors. Whenever a worker actor is reconstructed,
+RLlib will make sure its latest state gets synced before new episodes can be sampled. 
+
+RLlib workers are also elastic meaning that they do not require the full set of workers to
+function. For example, if an RLlib trial uses spot instances, it is quite possible for the
+trial to only have partial resources at a given time, potentially resulting in a subset
+of workers not being able to get scheduled. In this case, RLlib will make sure training and
```

Comment:
not being able to get scheduled -> not getting scheduled.

Reply:
done